### DISCUSSION TITLE
Hydrogen on Raspberry Pi 400

### DISCUSSION THREAD

**Discussion by dfkettle** [hydrogen-music/hydrogen#1479] [Hydrogen on Linux]
Hi!

I've just installed Hydrogen v1.0.1 on a Raspberry Pi 400 running Raspberry Pi OS. When I start Hydrogen, I get an error message "Error starting audio driver". So I went into Preferences and tried each of the options (Jack, ALSA, PortAudio and PulseAudio), but I get the same error for all of them. I'm using a Behringer UCA222 as my audio interface, which Audacity and Ardour both recognize, but Hydrogen doesn't seem to.

I know this isn't the latest version of Hydrogen, but it seems to be the latest that's available through the default respositories for the Raspberry Pi OS. Is this an issue with older versions of Hydrogen? I haven't tried downloading the latest version from Github, which I guess I'd have to compile myself. But if it will solve the problem, I'll give it a try.

Thanks.

**Comment by theGreatWhiteShark** [comment]
Hey @dfkettle,

> Is this an issue with older versions of Hydrogen?

Depends on the audio driver you are using. In terms of PortAudio there have been indeed a number of improvements.

Could you attach the debug log generated when starting Hydrogen in the terminal using `hydrogen -VDebug` and selecting the audio driver one after another in the Preferences? (in v1.0.1 you need at least to close the Preferences dialog in order for your selection to take effect. Only newer versions support restarting it right from within the dialog).

Could you also attach the output of the following commands `jack_control dp`, `pactl info`, and `aplay -L`?

**Comment by dfkettle** [comment]
Thanks. I'd forgotten that restarting the audio while you're still in Preferences doesn't work. When I selected ALSA and closed the Preferences window, it restarted without error. I don't know why the other three aren't working, but as long as one of them works, I don't really care.

**Reply by theGreatWhiteShark** [reply]
> I don't know why the other three aren't working, but as long as one of them works, I don't really care.

Maybe Pi OS has just a very minimal setup and ALSA is the only one supported.

But it still somehow feels like a bug. Does audio work if you select `Auto` in the Preferences? This should be the default selection on a fresh install and fall back to ALSA in case it is the only option available.

**Reply by dfkettle** [reply]
Yes, it defaults to Auto, but I then get the error message (which driver does "Auto" choose?). If I set it to ALSA, it's OK though.

If you still think it's a bug and not something I did wrong, I can send you the log files. I didn't want to waste your time, though.

**Reply by theGreatWhiteShark** [reply]
> which driver does "Auto" choose?

It tries all available audio drivers in a platform-dependent order until it finds one that is actually working. On Linux this is 
1. JACK
2. ALSA
3. OSS
4. PulseAudio
5. PortAudio

So, after not being able to use connect to the JACK server, it should have defaulted to the ALSA driver in your case.

It would be nice if you could provide the log file for starting Hydrogen with the Auto driver set. Then, I'll see whether I can reproduce it.

**Reply by dfkettle** [reply]
Here it is, along with the output from aplay and pactl. It seems jack_control isn't installed by default on the RPi, and I wasn't sure what package it's in (sudo apt-get install jack_control found nothing).

[aplay.log](https://github.com/hydrogen-music/hydrogen/files/8011323/aplay.log)
[hydrogen.log](https://github.com/hydrogen-music/hydrogen/files/8011324/hydrogen.log)
[pactl.log](https://github.com/hydrogen-music/hydrogen/files/8011325/pactl.log)

**Reply by theGreatWhiteShark** [reply]
> [hydrogen.log](https://github.com/hydrogen-music/hydrogen/files/8011324/hydrogen.log)

Hmm. That looks pretty good. The `Auto` option attempts to start the `JACK` driver but fails. Afterwards, it tries to start `ALSA` using device `hw:2`, which doesn't work either, but finally manages to start ALSA using the `default` device. Audio should work, does it?

Did you had to set some parameters of the "Alsa" option first in order for the `Auto` driver to succeed? Could you maybe produce a log file using the original settings for which no audio driver could be started? (e.g. by temporally moving your `~/.hydrogen/hydrogen.conf` file)

**Reply by theGreatWhiteShark** [reply]
> It seems jack_control isn't installed by default on the RPi, and I wasn't sure what package it's in (sudo apt-get install jack_control found nothing).

No problem. Seems JACK is not preinstalled on your system but this was most probably not the cause of the error.

**Reply by dfkettle** [reply]
That's what I uploaded, a log file using the original settings for which no audio driver could be started. I had to go into Preferences, select ALSA and then select hw:2 to get it to work. That's my USB audio interface. The default device is the HDMI port. If I set it back to Auto, it still doesn't work.

**Reply by theGreatWhiteShark** [reply]
That's weird.

But I just downgraded to version 1.0.1 and I got the same error as you did! Could you try [v1.1.1](https://github.com/hydrogen-music/hydrogen/releases/tag/1.1.1)? This might already be enough to fix it.

**Reply by dfkettle** [reply]
Yes, it did work! All options, including Auto (except for Jack), are working. Some day I might try installing Jack, but so far I haven't had any need for it, and it seems a little complicated to set up.

When you say you got the same error with v1.0.1, was that on a RPi? Or does the problem exist for all Linux devices?

BTW, I just downloaded the file 'hydrogen-1.1.1.tar.gz' and unzipped it. I'm not sure how to replace my existing install of v1.0.1 without compiling it from scratch. Is there an easy way to do that?

Thanks for your help!

**Reply by theGreatWhiteShark** [reply]
> When you say you got the same error with v1.0.1, was that on a RPi? Or does the problem exist for all Linux devices?

It seems to not be limited to RPis. I was using my regular laptop with all possible audio drivers being set up properly. Seems like we fixed more stuff during the past two years than I can recall :D

> BTW, I just downloaded the file 'hydrogen-1.1.1.tar.gz' and unzipped it. I'm not sure how to replace my existing install of v1.0.1 without compiling it from scratch. Is there an easy way to do that?

On Linux you, unfortunately, have to compile the code yourself. Only for macOS and Windows we ship binaries (mostly due to the abundance of Linux distributions).

> Is there an easy way to do that?

AFAIK Raspbian is based on Debian and in general it should be enough to copy&paste the code snippets from [INSTALL.md](https://github.com/hydrogen-music/hydrogen/blob/master/INSTALL.md)

**Reply by cme** [reply]
> AFAIK Raspbian is based on Debian and in general it should be enough to copy&paste the code snippets from [INSTALL.md](https://github.com/hydrogen-music/hydrogen/blob/master/INSTALL.md)

Can confirm -- I regularly build and run on a RPi 4 with Raspbian. There might be some missing package dependencies that I've installed that weren't in the list, but the method works!

**Reply by dfkettle** [reply]
> On Linux you, unfortunately, have to compile the code yourself. Only for macOS and Windows we ship binaries (mostly due to the abundance of Linux distributions).

But I didn't have to compile it. There was a binary in the tar file. Maybe I was just lucky because it seems to work.

**Reply by cme** [reply]
> But I didn't have to compile it. There was a binary in the tar file. Maybe I was just lucky because it seems to work.

...did a binary from my own raspi 4 escape into the tarball somehow? 🤣 I don't *think* so... that's very odd though.
